✨ Add push notifications support

sofokli1 opened the pull request:
The main idea is that we request a apnToken from the OS, send it to firebase so that we get in return an fcnToken. Send this token to our server so that we subscribe for push notifications. In addition there is a token storage to keep track of the last used token so that we do not do the requests if the apnToken hasn't changed.

I can send manual notifications using fcm token with the script below in bash
```
curl -X POST -H 'Content-type: application/json' -H "Authorization: key=FIREBASE_KEY" -d '{
  "registration_ids":["VERY_LONG_FCM_TOKEN_OF_YOUR_DEVICE_THAT_YOU_SHOULD_SEE_IN_THE_LOGS_WHEN_RUNNING_THE_APP"],
  "content_available": true
}' "https://fcm.googleapis.com/fcm/send"
```
you should see something in the log if you get a notification

There seems to be some issue for me getting notifications from our server, will investigate tomorrow with BE. But do check if it works for you. (it worked for me last week but not today)

The whole logic to handle the requests etc was added in the login action/reducer maybe it would make sense to have a separate action/reducer for the push notification tokens. Let me know what you think.

Odrakir commented on TogglTrack/Core/Actions/LoginAction.swift:
```diff
@@ -14,4 +14,7 @@ public enum LoginAction
     case logout
+    case subscribePushNotificationTokenToFCMIfNeeded(FCMPushToken)
+    case subscribedPushNotificationTokenToFCM(FCMResponse)
+    case subscribedPushNotificationTokenToToggl(FCMPushToken)
```
I don't think we need so many actions. Only one of them will ever be called from the app. The rest are always effects, so we don't really need to do the action-effect loop everytime. Maybe having two is enough (`subscribeToPushNotifications` and `subscribedToPushNotifications`), and the second one is just so we can print the second action and see it works.

sofokli1 replied:
not sure how drop the middle action right now, suggestions are welcome

Odrakir commented on TogglTrack/Model/FCMPushToken.swift:
```diff
@@ -0,0 +1,85 @@
+
+import Foundation
+
+public struct FCMPushToken: Codable, Equatable
```
I find this type really confusing, it seems to me like it's being used for too many things. Like, when we first get the device token maybe we just need to send a string to the reducer, and not wrap it in a type yet. 

Also, so many optional properties and the fact that coding and encoding keys are different... Maybe we need different types for sending requests and receiving responses.

Odrakir commented on TogglTrack/Environment/PushTokenStorage.swift:
```diff
@@ -0,0 +1,38 @@
+    func loadAPNToken() -> FCMPushToken?
+}
+
+public class PushTokenStorage: PushTokenStorageProtocol
```
I'm not sure if this service should be so specific. Maybe something wrapping user defaults is enough, and the specifics would be set from the reducer... You can leave it for now as we are not using user defaults anywhere else.

sofokli1 commented on TogglWatch WatchKit Extension/ExtensionDelegate.swift:
```diff
@@ -7,11 +7,15 @@
 
     func applicationDidFinishLaunching() {
-        // Perform any final initialization of your application.
+        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, error in
```
will test it

sofokli1 commented:
@Odrakir I am using string for the initial token we get from the OS instead of the model
- Changed the logic a bit and decided to always push the tokens (we never know when fcn invalidates the tokens)
- Added some logic to unsubscribe the token at logout
- Added a new appAction `.none` because I have an effect that does not need to go to any specific action (let me know what you think)
- I am open to suggestions on how to remove `subscribedPushNotificationTokenToFCM`, not sure now on how to do it at the moment.

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -67,3 +93,36 @@ private func loadUserEffect(_ api: APIProtocol) -> Effect<AppAction>
+        oldToken.apnToken != newAPNToken,
+        let oldTogglPushToken = oldToken.toTogglPushToken()
+    else {
+        return Just(.none).eraseToEffect()
```
```suggestion
        return .empty
```

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -67,3 +93,36 @@ private func loadUserEffect(_ api: APIProtocol) -> Effect<AppAction>
+        .map { _ in .none }
+        .catch { error in Just(.setError(error)) }
+        .eraseToEffect()
+}
```
Instead of the `.none` action how about creating this extension to `Effect`?

```swift
extension Publisher where Output == Void
{
    public func eraseToEmptyEffect<Action>(catch transformError: @escaping (Error) -> Action) -> Effect<Action>
    {
        self.flatMap { _ in
            Empty().eraseToAnyPublisher()
        }
        .catch({ error in Just(transformError(error))})
        .eraseToEffect()
    }
}
```

that way you can just do this and remove the `.none` action:

```swift
return api.unsubscribePushNotification(token: oldTogglPushToken)
    .eraseToEmptyEffect(catch: { error in .setError(error) })
```

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -41,14 +47,34 @@ public var loginReducer: Reducer<User?, LoginAction, LoginEnvironment, AppAction
+                .tags(.clear),
+                .timeline(.clear)
+            )
+        )
```
To delete the api token after all other effects we can create another extension on `Effect`:

```swift
extension Effect
{
    public static func fireAndForget(work: @escaping () -> Void) -> Effect
    {
        return Deferred { () -> Empty<Output, Never> in
            work()
            return Empty(completeImmediately: true)
        }
        .eraseToEffect()
    }
}
```

and call it:

```swift
Effect.fireAndForget {
    userEnv.keychain.deleteApiToken()
    userEnv.api.setAuth(token: nil)
}
```

sofokli1 commented:
@thanks for the help @Odrakir, please check again

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -67,3 +94,34 @@ private func loadUserEffect(_ api: APIProtocol) -> Effect<AppAction>
+private func subscribePushNotificationTokenToFCMEffect(_ firebaseAPI: FirebaseAPIProtocol, _ apnToken: String) -> Effect<AppAction>
+{
+    firebaseAPI.getFCMToken(for: FCMPushToken(apnToken))
+        .map { response in .user(.subscribedPushNotificationTokenToFCM(response)) }
```
To be able to remove the `subscribedPushNotificationTokenToFCM` action, instead of doing this, you could use `flatMap` and directly call `api.subscribePushNotification` here. Then you'll only need two actions: subscribe and unsubscribe.

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -67,3 +94,34 @@ private func loadUserEffect(_ api: APIProtocol) -> Effect<AppAction>
+
+private func subscribePushNotificationTokenToFCMEffect(_ firebaseAPI: FirebaseAPIProtocol, _ apnToken: String) -> Effect<AppAction>
+{
+    firebaseAPI.getFCMToken(for: FCMPushToken(apnToken))
```
I still think the `FCMPushToken` is confusing. Right here you are using a `FCMPushToken` to get a response of `[FCMPushToken]`, when in reality you only need to send the apn token, right? Or am I missing something?

sofokli1 replied:
apn token, sandbox flag, and app id. all of which the fcm token has and knows how to encode

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -103,10 +97,17 @@ private func subscribePushNotificationTokenToTogglEffect(_ api: APIProtocol, _ t
+        .tryFlatMap { (response) -> AnyPublisher<AppAction, Error> in
+            guard let fcmToken = response.fcmTokens?.first else { return Fail(error: UserError.NoFCMResponse).eraseToAnyPublisher() }
+            guard fcmToken.status == "OK" else { return Fail(error: UserError.FailedToGetFCMToken(fcmToken.status)).eraseToAnyPublisher() }
+            guard let togglToken = fcmToken.toTogglPushToken() else { return Fail(error: UserError.NoFCMTokenAvailable).eraseToAnyPublisher() }
```
On these three you can now just throw instead of returning a Fail Publisher, thanks to `tryFlatMap`

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -103,10 +97,17 @@ private func subscribePushNotificationTokenToTogglEffect(_ api: APIProtocol, _ t
 {
     firebaseAPI.getFCMToken(for: FCMPushToken(apnToken))
-        .map { response in .user(.subscribedPushNotificationTokenToFCM(response)) }
+        .tryFlatMap { (response) -> AnyPublisher<AppAction, Error> in
```
If you throw the errors you can get rid of the return type here and let type inference handle it.

Odrakir commented on TogglTrack/Core/Reducers/LoginReducer.swift:
```diff
@@ -69,3 +90,41 @@ private func loadUserEffect(_ api: APIProtocol) -> Effect<AppAction>
+        .catch { error in Just(.setError(error)) }
+        .eraseToEffect()
+}
+
```
🔥

Odrakir commented on TogglTrack/Core/Actions/LoginAction.swift:
```diff
@@ -14,4 +14,6 @@ public enum LoginAction
     case loadAPITokenAndUser
     case logout
+    case subscribePushNotificationTokenToFCM(String)
+    case subscribedPushNotificationTokenToToggl(FCMPushToken)
```
Now that we got rid of the other action, can't this be called something simpler like `subscribeToPushNotifications` and `unsubscribeToPushNotifications`?  Or maybe even move them to their own reducer and create `PushNotificationActions` `subscribe` and `unsubscribe`?

Odrakir commented on TogglWatch WatchKit Extension/ExtensionDelegate.swift:
```diff
@@ -57,4 +59,17 @@ class ExtensionDelegate: NSObject, WKExtensionDelegate {
+    {
+        guard let initialController = WKExtension.shared().rootInterfaceController as? HostingController else { return }
+        let deviceTokenString = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
+        initialController.store.send(.user(.subscribePushNotificationTokenToFCM(deviceTokenString)))
```
The action is now called `subscribeToPushNotifications`
